### PR TITLE
GL and helm version validation

### LINKED ISSUE #138: [HL] Use a Stipulated List of GitLab Versions
This is needed to ensure IaC compatibility, especially at major version release boundaries (e.g. 15.x broke, while 14.9 worked fine).

It would be best to only track Major and Minor because:
- tracking the patch (third segment) will be a pain because GitLab back patches old versions for security and so the list would need constant updating even though...
- it is unlikely that the changes in a patch release will affect compatibility of IaC

Seems like possibly using a pattern validation rule with comments about supported versions might be best?

### PR DESCRIPTION
**Pull request by DarwinJS**
*Issue #138 
Closes #138 

*Description of changes:*
Validates that GitLab and Helm versions have been tested with and are known to work with the QuickStart automation.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: templates/gitlab-entry-existing-vpc.template.yaml [hunk 1]
```diff
@@ -606,12 +606,25 @@ Parameters:
     Default: gitlab
   HelmChartVersion:
     Type: String
-    Description: Version of GitLab Helm chart GitLab for deployment. See https://docs.gitlab.com/charts/installation/version_mappings.html.
-    Default: '5.5.2'
+    AllowedPattern: '^(4\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d*|5\.(2|3|4|5|6|7|8|9|10)\.\d*)$'
```
Comment (tlindsay42) [inline templates/gitlab-entry-existing-vpc.template.yaml:609]: The `*`s should be swapped out with `+`s to mandate at least one digit for the patch version ID, right? For example, `14.10.` should not be an acceptable entry, right?

```diff
- ^(4\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d*|5\.(2|3|4|5|6|7|8|9|10)\.\d*)$
+ ^(4\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d+|5\.(2|3|4|5|6|7|8|9|10)\.\d+)$
```

File: templates/gitlab-entry-existing-vpc.template.yaml [hunk 2]
```diff
@@ -606,12 +606,25 @@ Parameters:
...
+    ConstraintDescription: |
+      Only GitLab versions 4.1-12 and 5.2-10 are currently tested and supported. 
+      The version format must be 'x.y.z' where all positions are numbers.
+    Description: |
+      Version of GitLab application - must correspond to helm chart version above. 
+      See https://docs.gitlab.com/charts/installation/version_mappings.html.
+    Default: '5.10.0'
   GitLabVersion:
     Type: String
+    AllowedPattern: ^(13\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d*|14\.(2|3|4|5|6|7|8|9|10)\.\d*)$
```
Comment (tlindsay42) [inline templates/gitlab-entry-existing-vpc.template.yaml:619]: Same as above

File: templates/gitlab-entry-new-vpc.template.yaml [hunk 3]
```diff
@@ -634,12 +634,29 @@ Parameters:
     Default: gitlab
   HelmChartVersion:
     Type: String
-    Description: Version of GitLab Helm chart GitLab for deployment. See https://docs.gitlab.com/charts/installation/version_mappings.html.
-    Default: '5.5.2'
+    AllowedPattern: '^(4\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d*|5\.(2|3|4|5|6|7|8|9|10)\.\d*)$'
```
Comment (tlindsay42) [inline templates/gitlab-entry-new-vpc.template.yaml:637]: Same as above

File: templates/gitlab-entry-new-vpc.template.yaml [hunk 4]
```diff
@@ -634,12 +634,29 @@ Parameters:
...
+      The version format must be 'x.y.z' where all positions are numbers.
+    Description: |
+      Version of GitLab application - must correspond to helm chart version above. 
+      See https://docs.gitlab.com/charts/installation/version_mappings.html.
+    Default: '5.10.0'
   GitLabVersion:
     Type: String
-    Description: Version of GitLab application - must correspond to helm chart version above. See https://docs.gitlab.com/charts/installation/version_mappings.html.
-    Default: '14.9.0'
+    AllowedPattern: '^(13\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d*|14\.(2|3|4|5|6|7|8|9|10)\.\d*)$'
```
Comment (tlindsay42) [inline templates/gitlab-entry-new-vpc.template.yaml:647]: Same as above

File: templates/gitlab-entry-existing-vpc.template.yaml [hunk 5]
```diff
@@ -606,12 +606,25 @@ Parameters:
...
+    ConstraintDescription: Only GitLab versions 13.1-12 and 14.2-10 are currently tested and supported. The version format must be 'x.y.z' where all positions are numbers.
     Description: Version of GitLab application - must correspond to helm chart version above. See https://docs.gitlab.com/charts/installation/version_mappings.html.
-    Default: '14.5.2'
+    Default: '14.10.0'
+    # While the AllowedPatterns for HelmChartVersion and GitLabVersion are verbose, they allows control over the
+    # maximum minor version including limiting it to 12 or less or completely eliminating minor versions in a series 
+    # For example, GitLab 14 only went to 14.10 and 14.1 did not work as CNH only on AWS due to classic load balancer.
+    # PATTERN LIKE: ^(13\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d*|14\.(1|2|3|4|5|6|7|8|9|10)\.\d*)$
```
Comment (tlindsay42) [inline templates/gitlab-entry-existing-vpc.template.yaml:626]: Same as above

File: templates/gitlab-entry-new-vpc.template.yaml [hunk 6]
```diff
@@ -634,12 +634,29 @@ Parameters:
...
+      Only GitLab versions 13.1-12 and 14.2-10 are currently tested and supported. 
+      The version format must be 'x.y.z' where all positions are numbers.
+    Description: |
+      Version of GitLab application - must correspond to helm chart version above. 
+      See https://docs.gitlab.com/charts/installation/version_mappings.html.
+    Default: '14.10.0'
+    # While the AllowedPatterns for HelmChartVersion and GitLabVersion are verbose, they allows control over the
+    # maximum minor version including limiting it to 12 or less or completely eliminating minor versions in a series 
+    # For example, GitLab 14 only went to 14.10 and 14.1 did not work as CNH only on AWS due to classic load balancer.
+    # PATTERN LIKE: ^(13\.(1|2|3|4|5|6|7|8|9|10|11|12)\.\d*|14\.(1|2|3|4|5|6|7|8|9|10)\.\d*)$
```
Comment (tlindsay42) [inline templates/gitlab-entry-new-vpc.template.yaml:658]: Same as above

File: templates/gitlab-entry-existing-vpc.template.yaml (hunk 1 above)
Comment (DarwinJS) [reply templates/gitlab-entry-existing-vpc.template.yaml]: Nice catch! Fixed

File: templates/gitlab-entry-existing-vpc.template.yaml (hunk 2 above)
Comment (DarwinJS) [reply templates/gitlab-entry-existing-vpc.template.yaml]: Nice catch! Fixed!

File: templates/gitlab-entry-new-vpc.template.yaml (hunk 3 above)
Comment (DarwinJS) [reply templates/gitlab-entry-new-vpc.template.yaml]: Nice catch! Fixed!

File: templates/gitlab-entry-existing-vpc.template.yaml (hunk 5 above)
Comment (DarwinJS) [reply templates/gitlab-entry-existing-vpc.template.yaml]: Nice catch! Fixed!

File: templates/gitlab-entry-new-vpc.template.yaml (hunk 6 above)
Comment (DarwinJS) [reply templates/gitlab-entry-new-vpc.template.yaml]: Nice catch! Fixed!

File: templates/gitlab-entry-new-vpc.template.yaml (hunk 4 above)
Comment (DarwinJS) [reply templates/gitlab-entry-new-vpc.template.yaml]: Nice catch! Fixed!

**Review by tlindsay42 — APPROVED** [review]

**Pull request merged**